Add compression support

wi-y opened the pull request:
Summary of changes

Normalization
- Add ability to compress the measurementGroup and send the compressed measurementGroup.
- Add way to flag if we are sending a measurement or measurementGroup (set IsMeasurementGroup property)
- Add way to flag if we are sending a compressed event payload (set ContentType to application/gzip)
- The compression on the normalization side is off by default, since we would like to first support it on the MeasurementToFhir side first

MeasurementToFhir
- Detect if event batch contains measurements, measurementGroups, or both
- Add ability to process any of the above, or combinations of the above
- Add ability to decompress event payload if it is compressed

rogordon01 commented on src/lib/Microsoft.Health.Fhir.Ingest/Service/MeasurementFhirImportService.cs:
```diff
@@ -90,9 +120,85 @@ public async Task ProcessEventsAsync(IEnumerable<IEventMessage> events, string t
+
+                        // decompress the measurement group if it is compressed
+                        byte[] bytes;
+                        if (evt.BodyContentType == "application/gzip")
```
If we go with my previous suggestions we could avoid extra object allocations here with something like:

```.net
                        IEnumerable<Measurement> measurementGroup = null;

                        if (evt.BodyContentType == "application/gzip")
                        {
                            using var stream = Compression.DecompressWithGzip(evt.Body.AsStream()); 
                            measurementGroup = System.Text.Json.JsonSerializer.Deserialize<IEnumerable<Measurement>>(stream);
                        }
                        else
                        {
                            using var stream = evt.Body.AsStream();
                            measurementGroup = System.Text.Json.JsonSerializer.Deserialize<IEnumerable<Measurement>>(stream)
                        }
```

dustinburson replied:
We should have constants for this.

wi-y commented on src/lib/Microsoft.Health.Fhir.Ingest/Service/MeasurementFhirImportService.cs:
```diff
@@ -90,9 +120,85 @@ public async Task ProcessEventsAsync(IEnumerable<IEventMessage> events, string t
+                        if (evt.BodyContentType == "application/gzip")
+                        {
+                            bytes = Compression.DecompressWithGzip(evt.Body.ToArray());
+                            evt.Body = new ReadOnlyMemory<byte>(bytes);
```
It is technically being used in the case where an exception is being thrown. If an exception is thrown I wanted the error message service to have access to the uncompressed bytes rather than the compressed bytes. But the downside is that we are setting this evt.Body on every message whether an exception is thrown or not. We could change this to have the error message service deal with the uncompressing if that is preferred

dustinburson replied:
I kind of like that approach but if we do that then we would want separate the responsibility for parsing the message into it's own class so we can reuse the logic.

wi-y replied:
I removed it for now and will add a work item for a refactor for the error message service

wi-y commented on src/lib/Microsoft.Health.Fhir.Ingest/Data/MeasurementToEventMessageAsyncCollector.cs:
```diff
@@ -23,15 +25,18 @@ public class MeasurementToEventMessageAsyncCollector :
             IHashCodeFactory hashCodeFactory,
-            ITelemetryLogger telemetryLogger)
+            ITelemetryLogger telemetryLogger,
+            IOptions<MeasurementToEventMessageAsyncCollectorOptions> options = null)
```
that makes sense

dustinburson commented on src/lib/Microsoft.Health.Fhir.Ingest/Service/MeasurementFhirImportService.cs:
```diff
@@ -90,9 +120,85 @@ public async Task ProcessEventsAsync(IEnumerable<IEventMessage> events, string t
+                        }
+
+                        // deserialize
+                        IEnumerable<Measurement> measurementGroup = JsonConvert.DeserializeObject<IEnumerable<Measurement>>(System.Text.Encoding.UTF8.GetString(bytes));
```
Is there an option with compression that we can get a stream instead of a byte array?  We can get the readonly memory stream from event body.  So if we can use a stream and instead of byte we should be able to avoid the array allocation for the byte array.

wi-y replied:
yeah we can use a stream. updated

dustinburson commented on src/lib/Microsoft.Health.Fhir.Ingest/Service/MeasurementFhirImportService.cs:
```diff
@@ -90,9 +120,85 @@ public async Task ProcessEventsAsync(IEnumerable<IEventMessage> events, string t
+                            bytes = Compression.DecompressWithGzip(evt.Body.ToArray());
+                            evt.Body = new ReadOnlyMemory<byte>(bytes);
+                        }
+                        else
```
Personally, I prefer using the switch expressions for scenarios like this.  See https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/switch-expression.  It doesn't really matter for one if/else statement but if we add other types the switch expression I feel results in more readable code compared to a large if/if else/else statement.

It would also be good to have a scenario if content is set to a type we don't support/know and raise an error

dustinburson reviewed (changes requested):
Overall changes look good. 

Two main things.  Where we are using byte[] if we can use a stream instead please look into making the switch.  This should reduce the memory allocation pressure we are seeing in perf tests.  Second, update the test cases so they also cover the compression scenarios.

Thanks!

wi-y commented on src/lib/Microsoft.Health.Common/IO/Compression.cs:
```diff
@@ -0,0 +1,42 @@
+                    }
+                }
+
+                return decompressedStream.ToArray();
```
thanks for the suggestion. updated

wi-y commented on src/lib/Microsoft.Health.Common/IO/Compression.cs:
```diff
@@ -0,0 +1,42 @@
+            }
+        }
+
+        public static byte[] DecompressWithGzip(byte[] bytes)
```
updated to use a stream

wi-y commented on test/Microsoft.Health.Events.UnitTest/EventBatchingTests.cs:
```diff
@@ -154,7 +154,7 @@ public async void GivenBatchServiceRunning_WhenNewEventConsumed_ThenCheckpointUp
 
             var firstEventTime = DateTime.UtcNow.AddSeconds(-301);
-            var firstEvent = new EventMessage("0", new ReadOnlyMemory<byte>(), 1, 1, firstEventTime, new Dictionary<string, object>(), new ReadOnlyDictionary<string, object>(new Dictionary<string, object>()));
+            var firstEvent = new EventMessage("0", new ReadOnlyMemory<byte>(), null, 1, 1, firstEventTime, new Dictionary<string, object>(), new ReadOnlyDictionary<string, object>(new Dictionary<string, object>()));
```
added some additional unit tests in MeasurementFhirImportServiceTests.cs to cover the new scenarios

dustinburson commented on test/Microsoft.Health.Common.UnitTests/Microsoft.Health.Common.UnitTests.csproj:
```diff
@@ -30,7 +30,6 @@
 			<PrivateAssets>all</PrivateAssets>
 			<IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
 		</PackageReference>
-		<PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
```
Reason for removal?  This was recently added because the Microsoft Test package has a Newtonsoft dependency and this ensures 13.0.1 or higher is used.

dustinburson commented on test/Microsoft.Health.Fhir.Ingest.UnitTests/Microsoft.Health.Fhir.Ingest.UnitTests.csproj:
```diff
@@ -18,7 +18,6 @@
       <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
     </PackageReference>
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.2.0" />
-    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
```
Same comment as before.  We should keep these references in the Test projects.

dustinburson commented on test/Microsoft.Health.Fhir.R4.Ingest.UnitTests/Microsoft.Health.Fhir.R4.Ingest.UnitTests.csproj:
```diff
@@ -18,7 +18,6 @@
       <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
     </PackageReference>
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.2.0" />
-    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
```
Same comment as before.  We should keep these references in the Test projects.

dustinburson commented on test/Microsoft.Health.Tests.Common/Microsoft.Health.Tests.Common.csproj:
```diff
@@ -27,7 +27,6 @@
       <PrivateAssets>all</PrivateAssets>
       <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
     </PackageReference>
-	<PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
```
Same comment as before.  We should keep these references in the Test projects.

dustinburson reviewed (changes requested):
Changes to use streams instead of byte[] arrays look good.

Only issue is the removal of the ```<PackageReference Include="Newtonsoft.Json" Version="13.0.1" />```  package references in the Test projects.  Please add them back in.  Please let me know if these were removed for a reason so we can see what options we have.  Thanks!

wi-y commented:
> Changes to use streams instead of byte[] arrays look good.
> 
> Only issue is the removal of the `<PackageReference Include="Newtonsoft.Json" Version="13.0.1" />` package references in the Test projects. Please add them back in. Please let me know if these were removed for a reason so we can see what options we have. Thanks!

I see. I've added the dependency back so that each test project has a Newtonsoft.Json 13.0.1 version.

There were two cases where there were duplicates, which caused a build error, so I have removed the duplicates in those two cases. All of the test csproj files should have exactly one Newtonsoft.Json now which I believe is the correct approach.

Cases where we had duplicates:
Microsoft.Health.Tests.Common.csproj
https://github.com/microsoft/iomt-fhir/blob/main/test/Microsoft.Health.Tests.Common/Microsoft.Health.Tests.Common.csproj#L30

Microsoft.Health.Fhir.Ingest.UnitTests.csproj
https://github.com/microsoft/iomt-fhir/blob/main/test/Microsoft.Health.Fhir.Ingest.UnitTests/Microsoft.Health.Fhir.Ingest.UnitTests.csproj#L34
